@W-14072135@ Add ability to wait for deployment to finish after pushing.

wjhsf opened the pull request:
<!--- Provide a short summary of your changes in the Title field above -->

Deploy as usual with `pwa-kit-dev push`, but add the `--wait` flag. The script will now wait for deployment to finish before exiting!

Now also includes a fix where `npm run push` didn't work in `template-retail-react-app`, because the output from `npm ls` is too big.

# Description

<!--- A longer summary of your changes, including: a description of the issue that you’re addressing, a list of required dependencies (if applicable), and any other relevant context. -->

# Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] **Bug fix** (non-breaking change that fixes an issue)
- [ ] **New feature** (non-breaking change that adds functionality)
- [ ] **Documentation update**
- [ ] **Breaking change** (could cause existing functionality to not work as expected)
- [ ] **Other changes** (non-breaking changes that does not fit any of the above)

> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

- (change1)

# How to Test-Drive This PR

- `npx pwa-kit-dev push --target some-target --wait`
- script will not exit until deployment completes!

# Checklists

<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [ ] Changes are covered by test cases
- [ ] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [ ] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [ ] Changes include a UI text update in the Retail React App (which requires translation)

wjhsf committed:
Add ability to wait for deployment to finish after pushing.

wjhsf committed:
Merge branch 'develop' into wjh/wait-for-deploy

wjhsf committed:
Throw an error if missing target when waiting.

wjhsf committed:
Add note to changelog.

shethj committed:
Merge branch 'develop' into wjh/wait-for-deploy

alexvuong commented:
Can you provide how to test drive this PR please?

vcua-mobify commented:
So the command to use this is `pwa-kit-dev push --target someTarget --wait` where target is an environment id from the project in MRT?

wjhsf committed:
Merge branch 'develop' into wjh/wait-for-deploy

wjhsf committed:
Clean up `readJson` mock.

wjhsf committed:
Add tests for `walkDir`.

wjhsf committed:
Correctly restore original `process.env`.

wjhsf commented on packages/pwa-kit-dev/src/ssr/server/build-dev-server.test.js:
```diff
@@ -267,16 +267,14 @@ describe('DevServer request processor support', () => {
+    let originalEnv = process.env
 
     beforeEach(() => {
-        originalEnv = Object.assign({}, process.env)
```
The original `process.env` object is never saved here, just cloned, so it gets polluted with test values and never reset.

wjhsf committed:
Use `--wait` in e2e CI.

wjhsf committed:
Make test platform agnostic.

wjhsf commented on packages/pwa-kit-dev/src/utils/script-utils.test.js:
```diff
@@ -488,4 +490,19 @@ describe('scriptUtils', () => {
         })
     })
+
+    describe('walkDir', () => {
```
Unrelated, needed to add tests to pass code coverage.

wjhsf committed:
Increase code coverage threshold.

Helps to avoid forgetting to add tests!

wjhsf committed:
Merge branch 'develop' into wjh/wait-for-deploy

vcua-mobify commented on packages/pwa-kit-dev/jest.config.js:
```diff
@@ -12,9 +12,9 @@ module.exports = {
-            statements: 70
+            functions: 75,
+            lines: 75,
+            statements: 75
```
I understand we'd like to have more coverage but could this threshold change be in a separate PR?

wjhsf replied:
If we don't uplift as we go, we're probably never going to get around to bumping the thresholds...

brianfeister replied:
I'm good to bump this in a PR. The key question in "should this be a different PR?" is "Will we create a new GUS ticket for this?" 

There will (in my opinion) always be a level of task granularity that falls to the side as "too small for a GUS ticket", I think it's acceptable here

vcua-mobify commented on packages/pwa-kit-dev/src/utils/script-utils.ts:
```diff
@@ -279,6 +279,65 @@ export class CloudAPIClient {
+    /** Polls MRT for deployment status every 60 seconds. */
+    async waitForDeploy(project: string, environment: string): Promise<void> {
+        return new Promise((resolve, reject) => {
+            const delay = 60e3 // 60 seconds
```
60 seconds is probably ok but it might also be a bit long (would 45 seconds be better, not sure)? Deploys that don't involve rebuilding the CDN don't take up a lot of time (probably around 2 min?)

wjhsf replied:
I've decreased the delay to 30 seconds.

bfeister replied:
This resolves a promise when the server indicated "DONE", right? So why are we worried about exiting early if the only chance of taking "too long" is that there is a real problem in the server layer?

vcua-mobify replied:
It's not so much as worried about exiting early but more not having to wait as long for the polling to say the build is complete once MRT is done. This shaves off 30 seconds of wait time for those who run the command locally and are not expecting a long build (as is the case when MRT skips rebuilding the CDN)

wjhsf committed:
Simplify logic with timeouts instead of interval.

wjhsf committed:
Reduce delay to 30 seconds.

bfeister commented on packages/pwa-kit-dev/src/utils/script-utils.ts:
```diff
@@ -279,6 +279,58 @@ export class CloudAPIClient {
+    async waitForDeploy(project: string, environment: string): Promise<void> {
+        return new Promise((resolve, reject) => {
+            /** Milliseconds to wait between checks. */
+            const delay = 30e3
```
Readability. I had to look up `30e3` I prefer to avoid code that makes you have to look things up unless it's necessary

vcua-mobify replied:
+1 on this. I had to look this up as well when I first saw it

bfeister commented on packages/pwa-kit-dev/src/utils/script-utils.ts:
```diff
@@ -279,6 +279,58 @@ export class CloudAPIClient {
+                        // Failed - reject with error
+                        return reject(new Error('Deployment failed.'))
+                    default:
+                        // Not in progress or failed - assume success
```
Why do we assume success if "not in progress or failed"? Sounds backwards

wjhsf replied:
🤷 I just copied the logic from the [existing code](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/1429/files#diff-d1c984792e22d7551aa2dbca33b9cb0014bf6bad55e4faa8474e391a7f188231). But looking at the back end code, it seems like the only other state a target can be is `ACTIVE`.

bfeister reviewed (commented):
Thanks for uplifting test coverage! 💪 Few questions / comments, no blockers

wjhsf committed:
Avoid exponential notation.

wjhsf committed:
Success is a known state.

wjhsf committed:
Use better error when project name exceeds max length.

wjhsf committed:
Fix `npm ls` on large projects.

`execSync` has a max buffer size smaller than the output of
`npm ls --all --json` for large projects.
Writing to a temp file avoids this issue.

wjhsf committed:
Maybe fix flaky test?

wjhsf commented on packages/pwa-kit-dev/src/ssr/server/build-dev-server.test.js:
```diff
@@ -304,10 +302,8 @@ describe('DevServer listening on http/https protocol', () => {
-                expect(response.ok).toBe(true)
-                return Promise.resolve()
-            })
+            const response = await insecureFetch(`${protocol}://localhost:${TEST_PORT}`)
+            expect(response.ok).toBe(true)
```
This test is flaky, so this is just an attempt to clean it up and maybe improve the flakiness (but probably not).

wjhsf committed:
Remove unnecessary mocks.

vcua-mobify reviewed (dismissed):
Going to approve this. I think it's looking good aside from some remaining questions and there's nothing blocking.

Let me know once we merge this in because I would like to have these changes in for the hybrid test environment deployment setup that I'm working on.

wjhsf committed:
Merge branch 'develop' into wjh/wait-for-deploy

wjhsf committed:
Increase test timeout.

bendvc commented on packages/pwa-kit-dev/src/utils/script-utils.test.js:
```diff
@@ -215,7 +226,27 @@ describe('scriptUtils', () => {
+                version: pkg.version,
+                dependencies: expect.any(Object)
+            })
+        }, 10_000) // This test can take a while on CI
```
C_O_O_L
